### PR TITLE
Workloads V1

### PR DESCRIPTION
**Pull request by kow3ns**
**What this PR does / why we need it**: This PR promotes the Deployment, ReplicaSet, and DaemonSet StatefulSet, ControllerRevision kinds to the apps/v1 group version.

https://github.com/kubernetes/features/issues/353

**Special notes for your reviewer**:
There will be at least two followups to this PR. The first to add a scale sub-resource when the correct location is resolved, and the second to deal with Conditions in the workloads API.

While it would have been preferable to move the kinds individually providing a lesser burden on reviewers, this proved impracticable due to the intricacies of version resolution in kubectl for objects of the different kinds in the same group.  

```release-note
DaemonSet, Deployment, ReplicaSet, and StatefulSet have been promoted to GA and are available in the apps/v1 group version.
```

### CODE REVIEW

**Comment by janetkuo** [comment]
> What this PR does / why we need it: This PR promotes the Deployment, ReplicaSet, and DaemonSet kinds to the apps/v1 group version.

PR title needs to be updated: replace DaemonSet with StatefulSet. Need to remove DS from release note too since it's already done.

**Comment by kow3ns** [comment]
StatefulSet is there, but we will probably do a single release note for the entire API.

**Comment by kow3ns** [comment]
@liggitt this version is without the scale
@kubernetes/sig-api-machinery-api-reviews 
@smarterclayton 
@bgrant0607

**Comment by bgrant0607** [comment]
@kow3ns I thought we were going to do this in a branch?

**Comment by kow3ns** [comment]
@bgrant0607 we are still trying to get the branch working and we didn't want to stop all development until then. The branch still can't run tests.

File: pkg/apis/apps/v1/defaults.go [hunk 1]
```diff
@@ -26,6 +26,49 @@ func addDefaultingFuncs(scheme *runtime.Scheme) error {
...
 }
 
+// SetDefaults_Deployment sets additional defaults compared to its counterpart
+// in extensions. These addons are:
+// - MaxUnavailable during rolling update set to 25% (1 in extensions)
+// - MaxSurge value during rolling update set to 25% (1 in extensions)
+// - RevisionHistoryLimit set to 10 (not set in extensions)
+// - ProgressDeadlineSeconds set to 600s (not set in extensions)
+func SetDefaults_Deployment(obj *appsv1.Deployment) {
+	// Set appsv1beta2.DeploymentSpec.Replicas to 1 if it is not set.
```
Comment (janetkuo) [inline pkg/apis/apps/v1/defaults.go]: nit: "appsv1"

File: hack/make-rules/test-cmd-util.sh [hunk 2]
```diff
@@ -1350,7 +1350,7 @@ run_kubectl_get_tests() {
   kube::test::if_has_string "${output_message}" "/api/v1/namespaces/default/pods 200 OK"
   kube::test::if_has_string "${output_message}" "/api/v1/namespaces/default/replicationcontrollers 200 OK"
   kube::test::if_has_string "${output_message}" "/api/v1/namespaces/default/services 200 OK"
-  kube::test::if_has_string "${output_message}" "/apis/apps/v1beta1/namespaces/default/statefulsets 200 OK"
+  kube::test::if_has_string "${output_message}" "/apis/apps/v1/namespaces/default/statefulsets 200 OK"
```
Comment (janetkuo) [inline hack/make-rules/test-cmd-util.sh]: Would this test be running against clusters that don't have `apps/v1` enabled yet?

**Comment by janetkuo** [comment]
Mention that `ControllerRevision` is bumped to apps/v1 in PR title/desc and release note too.

File: hack/make-rules/test-cmd-util.sh (hunk 2 above)
Comment (kow3ns) [reply hack/make-rules/test-cmd-util.sh]: The command tests are always written expecting a specific version of group version kind to be returned. That version has always been dependent on the packaged kubectl and apiserver configuration.

File: pkg/apis/apps/v1/defaults.go (hunk 1 above)
Comment (kow3ns) [reply pkg/apis/apps/v1/defaults.go]: Good catch. I'm going to rebase the update into the first commit on the PR.

File: pkg/kubectl/history.go [hunk 3]
```diff
@@ -52,26 +53,27 @@ type HistoryViewer interface {
 	ViewHistory(namespace, name string, revision int64) (string, error)
 }
```
Comment (janetkuo) [inline pkg/kubectl/history.go]: It looks like we don't need internal clientset at all in history viewer, do we?

File: pkg/kubectl/rollback.go [hunk 4]
```diff
@@ -54,20 +55,21 @@ type Rollbacker interface {
 	Rollback(obj runtime.Object, updatedAnnotations map[string]string, toRevision int64, dryRun bool) (string, error)
 }
```
Comment (janetkuo) [inline pkg/kubectl/rollback.go]: We probably don't need ic here either

File: pkg/kubectl/cmd/util/factory_object_mapping.go [hunk 5]
```diff
@@ -330,19 +330,21 @@ func (f *ring1Factory) Reaper(mapping *meta.RESTMapping) (kubectl.Reaper, error)
...
 	if err != nil {
 		return nil, err
 	}
-	return kubectl.HistoryViewerFor(mapping.GroupVersionKind.GroupKind(), clientset)
+	return kubectl.HistoryViewerFor(mapping.GroupVersionKind.GroupKind(), clientset, external)
 }
 
 func (f *ring1Factory) Rollbacker(mapping *meta.RESTMapping) (kubectl.Rollbacker, error) {
 	mappingVersion := mapping.GroupVersionKind.GroupVersion()
 	clientset, err := f.clientAccessFactory.ClientSetForVersion(&mappingVersion)
```
Comment (janetkuo) [inline pkg/kubectl/cmd/util/factory_object_mapping.go]: check `err`

File: pkg/kubectl/cmd/util/factory_object_mapping.go [hunk 6]
```diff
@@ -330,19 +330,21 @@ func (f *ring1Factory) Reaper(mapping *meta.RESTMapping) (kubectl.Reaper, error)
 func (f *ring1Factory) HistoryViewer(mapping *meta.RESTMapping) (kubectl.HistoryViewer, error) {
 	mappingVersion := mapping.GroupVersionKind.GroupVersion()
 	clientset, err := f.clientAccessFactory.ClientSetForVersion(&mappingVersion)
+	external, err := f.clientAccessFactory.KubernetesClientSet()
```
Comment (janetkuo) [inline pkg/kubectl/cmd/util/factory_object_mapping.go]: naming nit: if we need both internal and external clientset, name the internal one "internal" or name the external one "externalClientset"

File: hack/make-rules/test-cmd-util.sh (hunk 2 above)
Comment (janetkuo) [reply hack/make-rules/test-cmd-util.sh]: 👌

File: pkg/kubectl/history.go [hunk 7]
```diff
@@ -52,26 +53,27 @@ type HistoryViewer interface {
...
-		return &DaemonSetHistoryViewer{c}, nil
+		return &DaemonSetHistoryViewer{c, ec}, nil
 	}
 	return nil, fmt.Errorf("no history viewer has been implemented for %q", kind)
 }
 
 type DeploymentHistoryViewer struct {
-	c clientset.Interface
+	ic internalclientset.Interface
+	ec kubernetes.Interface
```
[the remark this answers is not in the document]
Comment (kow3ns) [reply pkg/kubectl/history.go]: We don't need any implementation detail in this context so  I think we should go with interface.

[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/kubectl/history.go]: +1 for interface

File: pkg/kubectl/history.go [hunk 8]
```diff
@@ -147,14 +149,15 @@ func printTemplate(template *v1.PodTemplateSpec) (string, error) {
 }
 
 type DaemonSetHistoryViewer struct {
-	c clientset.Interface
+	id internalclientset.Interface
```
[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/kubectl/history.go]: not sure how much is preexisting, but kubectl should be moving toward dealing in external versions
cc @kubernetes/sig-cli-pr-reviews

[the remark this answers is not in the document]
Comment (kow3ns) [reply pkg/kubectl/history.go]: I'm just going to go ahead and move the rollbackers and history to external version (v1beta1) and unwire the internal. We know the client is moving in this direction, and we might as well help the process along while there is time.

**Comment by kow3ns** [comment]
@janetkuo @liggitt I made the modification to remove the v1beta2 documentation reference in Defaults. I unwired the unversioned clientset as well. Both commits are squashed, the former into the api modification commit and the latter into the kubectl commit.

**Comment by mml** [comment]
@kow3ns is there anything specific you need api-machinery to review?

**Comment by kow3ns** [comment]
autocomplete did me wrong. I meant @kubernetes/api-reviewers

File: pkg/kubectl/history.go [hunk 9]
```diff
@@ -306,16 +251,59 @@ func controlledHistories(apps clientappsv1beta1.AppsV1beta1Interface, extensions
...
+		return nil, nil, fmt.Errorf("failed to obtain accessor for StatefulSet %s: %s", name, err.Error())
+	}
+	history, err := controlledHistory(apps, namespace, selector, accessor)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to find history controlled by StatefulSet %s: %v", name, err)
+	}
+	return sts, history, nil
+}
+
+// applyHistory returns a specific revision of DaemonSet by applying the given history to a copy of the given DaemonSet
```
Comment (janetkuo) [inline pkg/kubectl/history.go]: nit: applyDaemonSetHistory

File: pkg/kubectl/rollback.go [hunk 10]
```diff
@@ -212,48 +213,40 @@ func simpleDryRun(deployment *extensions.Deployment, c clientset.Interface, toRe
...
 	if err != nil {
-		return "", fmt.Errorf("unable to find history controlled by DaemonSet %s: %v", ds.Name, err)
+		return "", fmt.Errorf("failed to create accessor for kind %v: %s", obj.GetObjectKind(), err.Error())
 	}
-	versionedDS, ok := versionedObj.(*externalextensions.DaemonSet)
-	if !ok {
-		return "", fmt.Errorf("unexpected non-DaemonSet object returned: %v", versionedDS)
+	ds, history, err := daemonSetHistory(r.c.ExtensionsV1beta1(), r.c.AppsV1beta1(), accessor.GetNamespace(), accessor.GetName())
+	if err != nil {
+		return "", fmt.Errorf("unable to find history controlled by DaemonSet %s: %s", ds.Name, err.Error())
```
Comment (janetkuo) [inline pkg/kubectl/rollback.go]: Just return `err`

File: pkg/kubectl/rollback.go [hunk 11]
```diff
@@ -262,58 +255,49 @@ func (r *DaemonSetRollbacker) Rollback(obj runtime.Object, updatedAnnotations ma
...
-		return "", fmt.Errorf("unable to find history controlled by StatefulSet %s: %v", ss.Name, err)
+		return "", fmt.Errorf("failed to create accessor for kind %v: %s", obj.GetObjectKind(), err.Error())
 	}
-
-	versionedSS, ok := versionedObj.(*appsv1beta1.StatefulSet)
-	if !ok {
-		return "", fmt.Errorf("unexpected non-StatefulSet object returned: %v", versionedSS)
+	sts, history, err := statefulSetHistory(r.c.AppsV1beta1(), accessor.GetNamespace(), accessor.GetName())
+	if err != nil {
+		return "", fmt.Errorf("unable to find history controlled by StatefulSet %s: %s", accessor.GetName(), err.Error())
```
Comment (janetkuo) [inline pkg/kubectl/rollback.go]: simply return `err`

**Comment by DirectXMan12** [comment]
> There will be at least two followups to this PR. The first to add a scale sub-resource when the correct location is resolved

NB: there should never be a `Scale` type in `apps/v1`.  The only scale types should be in `autoscaling/v1` and `extensions/v1beta1`.  Newer resources should return the `autoscaling/v1` Kind from their `/scale` subresource (see https://github.com/kubernetes/kubernetes/issues/49504#issuecomment-336020218)

**Comment by kow3ns** [comment]
@DirectXMan12 We need to register a scale subresource in the apps group. It doesn't say that we will add a Scale resource in apps/v1.

**Comment by DirectXMan12** [comment]
Yep, just making sure the mistake of having a `Scale` type in apps/v1beta2 isn't repeated in apps/v1.

**Review by janetkuo — APPROVED** [review]

**Comment by janetkuo** [comment]
/approve

I added https://github.com/kubernetes/features/issues/353 to the PR description.

**Comment by smarterclayton** [comment]
>  this proved impracticable due to the intricacies of version resolution in kubectl for objects of the different kinds in the same group.

Is this recorded as an issue somewhere?

**Comment by kow3ns** [comment]
@smarterclayton The general approach we are going to take with kubectl is to move to the versioned API. This will handle correct content negotiation across versions using the same kubernetes.Interface. I'm not sure if this particular facet of the problems we have faced using inteternal_clientset.Interface have been documented.

**Comment by smarterclayton** [comment]
We do version negotiation all of the time in kubectl so I want to understand the specific thing you are mentioning (which sounds like a bug).

There are many types of commands that will be used in kubectl and extensions that will continue to use internal versions.  Even if our approach is to mostly move, it does not remove the need for kubectl rest mappers to deal with version resolution for generic clients.  create still has to do the proper negotiation for types.

**Comment by kow3ns** [comment]
It seems that once the client uses the apps/v1 group it defaults to that version. Attempt to retrieve an object using the same client instance and at lower version v1beta2 result in a request and response against the v1 version rather than the v1beta2 version.

**Comment by liggitt** [comment]
> It seems that once the client uses the apps/v1 group it defaults to that version. Attempt to retrieve an object using the same client instance and at lower version v1beta2 result in a request and response against the v1 version rather than the v1beta2 version.

do you have an example of the client calls you're referring to? `kubectl get ...`? or `client.AppsV1beta2().ReplicaSets("foo").Get(...)` or something else?

**Comment by kow3ns** [comment]
If you have sts and rs in different versions of the same group and you perform
Apps().StatefulSets().Get(...)
subsequent calls to Apps().ReplicaSets().Get(...) fail to find the requested resource. The do not fall back v1beta1.
When we moved over each controller piecemeal in separate PRs the command tests failed. DaemonSet appears to work correctly, and independently, because kubectl targets the extensions group rather than apps.

**Comment by liggitt** [comment]
> Apps().ReplicaSets().Get(...) fail to find the requested resource. The do not fall back v1beta1.

Apps() has never done version negotiation or fallback. It is an alias for one of the versioned client accessors (documented deprecated). Uses should be converted to a specific version. 

See https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/kubernetes/clientset.go

**Comment by kow3ns** [comment]
internal_clientset not kubernetes

File: pkg/registry/extensions/daemonset/strategy.go [hunk 12]
```diff
@@ -135,7 +136,7 @@ func (daemonSetStrategy) ValidateUpdate(ctx genericapirequest.Context, obj, old
 		switch groupVersion {
 		case extensionsv1beta1.SchemeGroupVersion:
 			// no-op for compatibility
-		case appsv1beta2.SchemeGroupVersion:
+		case appsv1beta2.SchemeGroupVersion, appsv1.SchemeGroupVersion:
```
Comment (liggitt) [inline pkg/registry/extensions/daemonset/strategy.go]: add strategy test that exercises this

File: pkg/registry/extensions/deployment/strategy.go [hunk 13]
```diff
@@ -117,7 +118,7 @@ func (deploymentStrategy) ValidateUpdate(ctx genericapirequest.Context, obj, old
 			// no-op for compatibility
 		case extensionsv1beta1.SchemeGroupVersion:
 			// no-op for compatibility
-		case appsv1beta2.SchemeGroupVersion:
+		case appsv1beta2.SchemeGroupVersion, appsv1.SchemeGroupVersion:
```
Comment (liggitt) [inline pkg/registry/extensions/deployment/strategy.go]: add strategy test that exercises this

File: pkg/registry/extensions/replicaset/strategy.go [hunk 14]
```diff
@@ -127,7 +128,7 @@ func (rsStrategy) ValidateUpdate(ctx genericapirequest.Context, obj, old runtime
 		switch groupVersion {
 		case extensionsv1beta1.SchemeGroupVersion:
 			// no-op for compatibility
-		case appsv1beta2.SchemeGroupVersion:
+		case appsv1beta2.SchemeGroupVersion, appsv1.SchemeGroupVersion:
```
Comment (liggitt) [inline pkg/registry/extensions/replicaset/strategy.go]: add strategy test that exercises this

**Comment by liggitt** [comment]
type and history client changes LGTM. were there any changes in the non-generated defaulting/conversion?

**Comment by kow3ns** [comment]
@liggitt  defaults/conversions remain the same as apps/v1beta2. This promotes what was discussed last release minus the scale sub-resource. I added the tests in into the first commit on the PR so that they went on top of the auto-gen.

**Review by janetkuo — APPROVED** [review]

**Comment by krmayankk** [comment]
@kow3ns what would be needed for users to migrate from v1beta1 to v1  for these objects ?

**Comment by kow3ns** [comment]
@krmayankk the storage formats are not changed. You can consume the API from apps/v1. The deprecation cycle for extensions/v1, apps/v1beta1, and apps/v1beta2 will take into account that users are currently depending on them in production. Subsequent releases will include instructions on how to upgrade storage formats in order to migrate existing workloads to a newer version.

File: test/integration/etcd/etcd_storage_path_test.go [hunk 15]
```diff
@@ -181,6 +181,26 @@ var etcdStorageData = map[schema.GroupVersionResource]struct {
...
+		expectedEtcdPath: "/registry/replicasets/etcdstoragepathtestnamespace/rs3",
+		expectedGVK:      gvkP("extensions", "v1beta1", "ReplicaSet"),
+	},
+	gvr("apps", "v1", "controllerrevisions"): {
+		stub:             `{"metadata":{"name":"crs3"},"data":{"name":"abc","namespace":"default","creationTimestamp":null,"Spec":{"Replicas":0,"Selector":{"matchLabels":{"foo":"bar"}},"Template":{"creationTimestamp":null,"labels":{"foo":"bar"},"Spec":{"Volumes":null,"InitContainers":null,"Containers":null,"RestartPolicy":"Always","TerminationGracePeriodSeconds":null,"ActiveDeadlineSeconds":null,"DNSPolicy":"ClusterFirst","NodeSelector":null,"ServiceAccountName":"","AutomountServiceAccountToken":null,"NodeName":"","SecurityContext":null,"ImagePullSecrets":null,"Hostname":"","Subdomain":"","Affinity":null,"SchedulerName":"","Tolerations":null,"HostAliases":null}},"VolumeClaimTemplates":null,"ServiceName":""},"Status":{"ObservedGeneration":null,"Replicas":0}},"revision":0}`,
```
Comment (liggitt) [inline test/integration/etcd/etcd_storage_path_test.go]: this stub is using JSON generated from an internal object (note the capitalized keys). this is pre-existing in the other stub tests and can be fixed as a follow up.

**Comment by pwittrock** [comment]
/approve

approve for hack

**Pull request merged**